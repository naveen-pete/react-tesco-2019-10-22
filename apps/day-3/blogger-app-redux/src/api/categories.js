const apiUrl = 'http://localhost:3001/categories';

export const getCategories = () => {
  return fetch(apiUrl)
    .then(response => response.json());
};

export const getCategory = id => {
  return fetch(`${apiUrl}/${id}`)
    .then(response => response.json());
};
