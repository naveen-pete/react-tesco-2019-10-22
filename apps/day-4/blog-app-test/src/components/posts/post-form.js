import React, { Component } from 'react';
import FontAwesome from 'react-fontawesome';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';

import {
  getCategories,
  getPost,
  createPost,
  updatePost
} from '../../actions/posts';

const defaultPost = {
  title: '',
  body: '',
  author: '',
  category: '',
  commentCount: 0,
  voteScore: 0,
  timestamp: 0
};

export class PostForm extends Component {
  constructor() {
    super();

    this.state = {
      post: defaultPost
    };

    this.onFormSubmit = this.onFormSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    this.props.getCategories();
    const id = this.props.match.params.id;
    if (id) {
      this.props.getPost(id, post => {
        this.setState({ post });
      });
    }
  }

  onFormSubmit(event) {
    event.preventDefault();

    // #1: older style - using Object.assign() method
    // const post = Object.assign({}, this.state.post);

    // #2: newer style - using object destructuring syntax
    const post = { ...this.state.post };
    const callback = () => this.props.history.push('/');

    if (post.id) {
      this.props.updatePost(post, callback);
    } else {
      post.timestamp = Date.now();
      this.props.createPost(post, callback);
    }
  }

  handleChange(event) {
    // Retrieve name, value and control type flag of the form control
    const { name, value } = event.target;

    this.setState(prevState => {
      // Assign the updated value
      // #1: older style - using Object.assign() method
      // const newPost = Object.assign({}, prevState.post, { [name]: value });

      // #2: newer style - using object destructuring syntax
      const newPost = {
        ...prevState.post,
        [name]: value
      };

      return { post: newPost };
    });
  }

  render() {
    const { post } = this.state;
    const { categories } = this.props;

    return (
      <div className="row">
        <div className="col-sm-12">
          <h4>
            <FontAwesome name="envelope" /> Post Form
          </h4>
          <div className="well well-lg">
            <form onSubmit={this.onFormSubmit}>
              <div className="form-group">
                <label htmlFor="title">Title</label>
                <input
                  required
                  type="text"
                  className="form-control"
                  id="title"
                  name="title"
                  value={post.title}
                  onChange={this.handleChange}
                />
              </div>

              <div className="form-group">
                <label htmlFor="body">Body</label>
                <textarea
                  required
                  className="form-control"
                  id="body"
                  name="body"
                  cols="30"
                  rows="4"
                  value={post.body}
                  onChange={this.handleChange}
                />
              </div>

              <div className="form-group">
                <label htmlFor="author">Author</label>
                <input
                  required
                  type="text"
                  className="form-control"
                  id="author"
                  name="author"
                  value={post.author}
                  onChange={this.handleChange}
                />
              </div>

              <div className="form-group">
                <label htmlFor="category">Category</label>
                <select
                  required
                  className="form-control"
                  id="category"
                  name="category"
                  value={post.category}
                  onChange={this.handleChange}
                >
                  <option value="" />
                  {categories.map(({ code, name }) => (
                    <option key={code} value={code}>
                      {name}
                    </option>
                  ))}
                </select>
              </div>

              <button
                type="submit"
                className="btn btn-primary btn-sm smallMargin"
              >
                <span className="glyphicon glyphicon-ok" /> Save
              </button>
              <Link to="/" className="btn btn-warning btn-sm smallMargin">
                <span className="glyphicon glyphicon-remove" /> Cancel
              </Link>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ categories }, props) => ({
  categories
});

const mapDispatchToProps = dispatch => ({
  getCategories: () => dispatch(getCategories()),
  getPost: (id, callback) => dispatch(getPost(id, callback)),
  createPost: (post, callback) => dispatch(createPost(post, callback)),
  updatePost: (post, callback) => dispatch(updatePost(post, callback))
});

export default connect(mapStateToProps, mapDispatchToProps)(PostForm);
