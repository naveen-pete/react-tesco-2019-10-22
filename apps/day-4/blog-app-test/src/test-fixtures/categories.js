const categories = [
  {
    "name": "React",
    "code": "react"
  },
  {
    "name": "Redux",
    "code": "redux"
  },
  {
    "name": "Angular",
    "code": "angular"
  },
  {
    "name": "ES6",
    "code": "es6"
  }
];

export default categories;
