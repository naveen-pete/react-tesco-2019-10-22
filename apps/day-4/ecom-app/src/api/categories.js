const apiUrl = 'http://localhost:3000/api/categories';

export const getCategories = async () => {
  const response = await fetch(apiUrl);
  return response.json();
};
