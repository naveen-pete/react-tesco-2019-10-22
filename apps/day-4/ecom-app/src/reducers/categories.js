
const categoriesReducer = (state = [], action) => {
  switch (action.type) {
    case 'GET_CATEGORIES':
      return [...action.payload];

    default:
      return state;
  };
};

export default categoriesReducer;
