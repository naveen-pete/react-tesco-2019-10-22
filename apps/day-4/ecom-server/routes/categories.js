const express = require('express');
const router = express.Router();

const { Category } = require('../models/category');

router.get('/', async (req, res) => {
  try {
    const categories = await Category.find().select({ __v: 0 });
    res.json(categories);
  } catch (e) {
    res.status(500).json({ message: 'Server error.' });
    console.log('Error:', e.message);
  }
});

router.get('/:id', async (req, res) => {
  try {
    const category = await Category.findById(req.params.id).select({ __v: 0 });

    if (!category) {
      return res.status(404).json({ message: 'Category does not exist!' });
    }

    res.json(category);
  } catch (e) {
    res.status(500).json({ message: 'Server error.' });
    console.log('Error:', e.message);
  }
});

module.exports = router;