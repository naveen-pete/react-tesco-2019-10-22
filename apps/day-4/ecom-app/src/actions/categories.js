import { getCategories as getCategoriesApi } from '../api/categories';

export const getCategories = () => {
  return async dispatch => {
    try {
      const categories = await getCategoriesApi();
      dispatch({
        type: 'GET_CATEGORIES',
        payload: categories
      });
    } catch (e) {
      console.log('Get categories failed!')
      console.log('Error:', e);
    }
  };
}