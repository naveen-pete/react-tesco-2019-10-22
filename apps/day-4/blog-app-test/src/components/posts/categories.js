import React, { Component } from 'react';
import FontAwesome from 'react-fontawesome';

import { connect } from 'react-redux';

import { CATEGORY_ALL } from './defaults';
import { getCategories, setCurrentCategory } from '../../actions/posts';

export class Categories extends Component {
  componentDidMount() {
    this.props.getCategories();
  }

  onCategoryClick(selectedCategory) {
    this.props.setCurrentCategory(selectedCategory);
  }

  addActiveIfSelected(category) {
    const { code } = this.props.currentCategory;
    return code === category.code ? 'active' : '';
  }

  render() {
    let { categories } = this.props;
    categories = [CATEGORY_ALL, ...categories];

    return (
      <div className="col-sm-3">
        <h4>
          <FontAwesome name="object-group" /> Category
        </h4>
        <ul className="list-group">
          {categories.map(category => (
            <li
              onClick={() => this.onCategoryClick(category)}
              key={category.code}
              className={`list-group-item cursorPointer ${this.addActiveIfSelected(category)}`}
            >
              {category.name}
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

const mapStateToProps = ({ categories, currentCategory }) => ({
  categories,
  currentCategory
});

const mapDispatchToProps = dispatch => ({
  getCategories: () => dispatch(getCategories()),
  setCurrentCategory: category => dispatch(setCurrentCategory(category))
});

export default connect(mapStateToProps, mapDispatchToProps)(Categories);
