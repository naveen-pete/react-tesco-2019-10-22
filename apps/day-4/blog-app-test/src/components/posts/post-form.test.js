import React from 'react';
import { shallow } from 'enzyme';
import { PostForm } from './post-form';
import categories from '../../test-fixtures/categories';

let realDateNow, mockDateNow, match, defaultPost, getCategoriesSpy, 
  getPostSpy, createPostSpy, updatePostSpy;

beforeEach(() => {
  match = {
    params: { id: undefined }
  };

  defaultPost = {
    title: '',
    body: '',
    author: '',
    category: '',
    commentCount: 0,
    voteScore: 0,
    timestamp: 0
  };

  getCategoriesSpy = jest.fn();
  getPostSpy = jest.fn();
  createPostSpy = jest.fn();
  updatePostSpy = jest.fn();

  realDateNow = Date.now;
  mockDateNow = jest.fn().mockImplementation(() => 123);
  Date.now = mockDateNow;
});

afterEach(() => {
  Date.now = realDateNow;
});

test('should render the form', () => {
  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getPost={getPostSpy} 
      getCategories={getCategoriesSpy} 
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('should populate categories within Category dropdown list', () => {
  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getPost={getPostSpy} 
      getCategories={getCategoriesSpy} 
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('#category > option').length).toBe(categories.length + 1);
});

test('should call getCategories() when the component is mounted', () => {
  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getPost={getPostSpy} 
      getCategories={getCategoriesSpy} 
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(getCategoriesSpy).toHaveBeenCalled();
});

test('should set state to default post when the component is mounted', () => {
  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getPost={getPostSpy} 
      getCategories={getCategoriesSpy} 
    />
  );
  expect(wrapper).toMatchSnapshot();
  // #1
  // expect(wrapper.state('post')).toEqual(defaultPost);

  // #2
  expect(wrapper.state().post).toEqual(defaultPost);
});

test('should call getPost() when the component is mounted and post id is passed as a parameter', () => {
  const id = 1;
  match.params.id = id;

  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getPost={getPostSpy} 
      getCategories={getCategoriesSpy} 
    />
  );
  expect(wrapper).toMatchSnapshot();
  expect(getCategoriesSpy).toHaveBeenCalled();
  expect(getPostSpy).toHaveBeenCalledWith(id, expect.any(Function));
});

test('should save post title in state when the value in title text box is changed', () => {
  const name = 'title';
  const value = 'my test post';

  const event = { 
    target: { name, value }
  };

  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getPost={getPostSpy} 
      getCategories={getCategoriesSpy} 
    />
  );

  expect(wrapper).toMatchSnapshot();

  wrapper.find('#title').simulate('change', event);
  const statePost =  wrapper.state('post');
  expect(statePost.title).toBe(value);
});

test('should save post body in state when the value in body text area is changed', () => {
  const name = 'body';
  const value = 'my test post body';

  const event = { 
    target: { name, value }
  };

  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getPost={getPostSpy} 
      getCategories={getCategoriesSpy} 
    />
  );

  expect(wrapper).toMatchSnapshot();

  wrapper.find('#body').simulate('change', event);
  const statePost =  wrapper.state('post');
  expect(statePost.body).toBe(value);
});

test('should save selected category in state when a category is selected', () => {
  const name = 'category';
  const value = 'testCat';

  const event = { 
    target: { name, value }
  };

  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getPost={getPostSpy} 
      getCategories={getCategoriesSpy} 
    />
  );

  expect(wrapper).toMatchSnapshot();

  wrapper.find('#category').simulate('change', event);
  const statePost =  wrapper.state('post');
  expect(statePost.category).toBe(value);
});

test('should add a post when the form is submitted', () => {
  const post = {
    title: 'test post',
    body: 'test post body',
    author: 'test author',
    category: 'testCat',
    commentCount: 1,
    voteScore: 2,
    timestamp: 123
  };

  const event = { 
    preventDefault: jest.fn() 
  };

  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getCategories={getCategoriesSpy} 
      createPost={createPostSpy}
    />
  );

  expect(wrapper).toMatchSnapshot();

  wrapper.setState({post}, () => {
    wrapper.find('form').simulate('submit', event);
    expect(event.preventDefault).toHaveBeenCalled();
    expect(mockDateNow).toHaveBeenCalled();
    expect(createPostSpy).toHaveBeenCalledWith(post, expect.any(Function));
  });
});

test('should update a post when the form is submitted', () => {
  const post = {
    id: 2,
    title: 'test post',
    body: 'test post body',
    author: 'test author',
    category: 'testCat',
    commentCount: 1,
    voteScore: 2,
    timestamp: 123
  };

  const event = { 
    preventDefault: jest.fn() 
  };

  const wrapper = shallow(
    <PostForm 
      match={match}
      categories={categories}
      getCategories={getCategoriesSpy} 
      updatePost={updatePostSpy}
    />
  );

  expect(wrapper).toMatchSnapshot();

  wrapper.setState({post}, () => {
    wrapper.find('form').simulate('submit', event);
    expect(event.preventDefault).toHaveBeenCalled();
    expect(updatePostSpy).toHaveBeenCalledWith(post, expect.any(Function));
  });
});
