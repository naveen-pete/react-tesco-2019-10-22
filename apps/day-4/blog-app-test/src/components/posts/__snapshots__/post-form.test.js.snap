// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a post when the form is submitted 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should call getCategories() when the component is mounted 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should call getPost() when the component is mounted and post id is passed as a parameter 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should populate categories within Category dropdown list 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should render the form 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should save post body in state when the value in body text area is changed 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should save post title in state when the value in title text box is changed 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should save selected category in state when a category is selected 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should set state to default post when the component is mounted 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;

exports[`should update a post when the form is submitted 1`] = `
<div
  className="row"
>
  <div
    className="col-sm-12"
  >
    <h4>
      <FontAwesome
        name="envelope"
      />
       Post Form
    </h4>
    <div
      className="well well-lg"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="title"
          >
            Title
          </label>
          <input
            className="form-control"
            id="title"
            name="title"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="body"
          >
            Body
          </label>
          <textarea
            className="form-control"
            cols="30"
            id="body"
            name="body"
            onChange={[Function]}
            required={true}
            rows="4"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="author"
          >
            Author
          </label>
          <input
            className="form-control"
            id="author"
            name="author"
            onChange={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="category"
          >
            Category
          </label>
          <select
            className="form-control"
            id="category"
            name="category"
            onChange={[Function]}
            required={true}
            value=""
          >
            <option
              value=""
            />
            <option
              key="react"
              value="react"
            >
              React
            </option>
            <option
              key="redux"
              value="redux"
            >
              Redux
            </option>
            <option
              key="angular"
              value="angular"
            >
              Angular
            </option>
            <option
              key="es6"
              value="es6"
            >
              ES6
            </option>
          </select>
        </div>
        <button
          className="btn btn-primary btn-sm smallMargin"
          type="submit"
        >
          <span
            className="glyphicon glyphicon-ok"
          />
           Save
        </button>
        <Link
          className="btn btn-warning btn-sm smallMargin"
          replace={false}
          to="/"
        >
          <span
            className="glyphicon glyphicon-remove"
          />
           Cancel
        </Link>
      </form>
    </div>
  </div>
</div>
`;
